plugin/view/README.md

```diff
@@ -0,0 +1,117 @@
+# view
+
+## Name 
+
+*view* - defines conditions that must be met for a DNS request to be routed to the server block.
+
+## Description
+
+*view* defines an expression that must evaluate to true for a DNS request to be routed to the server block.  
+This enables advanced server block routing functions such as split dns.    
+
+## Syntax
+```
+view  EXPRESSION
+```
+
+* `view` **EXPRESSION** - CoreDNS will only route incoming queries to the enclosing server block
+  if the **EXPRESSION** evaluates to true. See the **Expressions** section for available variables and functions.
+  If multiple instances of view are defined, all **EXPRESSION** must evaluate to true for CoreDNS will only route
+  incoming queries to the enclosing server block.
+
+For expression syntax and examples, see the Expressions and Examples sections.
+
+## Examples
+
+Implement CIDR based split DNS routing.  This will return a different
+answer for `test.` depending on client's IP address.  It returns ...
+* `test. 3600 IN A 1.1.1.1`, for queries with a source address in 127.0.0.0/24
+* `test. 3600 IN A 2.2.2.2`, for queries with a source address in 192.168.0.0/16
+* `test. 3600 IN A 3.3.3.3`, for all others
+
+```
+. {
+  view incidr(client_ip, '127.0.0.0/24')
+  hosts {
+    1.1.1.1 test
+  }
+}
+
+. {
+  view incidr(client_ip, '192.168.0.0/16')
+  hosts {
+    2.2.2.2 test
+  }
+}
+
+. {
+  hosts {
+    3.3.3.3 test
+  }
+}
+```
+
+Send all `AAAA` requests to `10.0.0.6`, and all other requests to `10.0.0.1`.
+
+```
+. {
+  view type == 'AAAA'
+  forward . 10.0.0.6
+}
+
+. {
+  forward . 10.0.0.1
+}
+```
+
+Send all requests for `abc.*.example.com` (where * can be any number of labels), to `10.0.0.2`, and all other
+requests to `10.0.0.1`.
+
+```
+. {
+  view name =~ '^abc\..*\.example\.com\.$'
+  forward . 10.0.0.2
+}
+
+. {
+  forward . 10.0.0.1
+}
+```
+
+## Expressions
+
+Expressions use Kinetic.govaluate (https://github.com/Knetic/govaluate), which "Provides support for evaluating arbitrary
```

Comment:
I would be quite hesitant to add a new logic on top of [govaluate](https://github.com/Knetic/govaluate) project, it seems to be no longer actively maintained. Is there any alternative we can use? Would it be possible for example to use standard GO templates (like in [Helm](https://helm.sh/)) for this expression logic?

Reply:
Googling for "govaluate alternatives" First hit is https://github.com/antonmedv/expr, which looks promising, and appears actively maintained.

Reply:
that seems better choice than govaluate, if you think that GO templates are not suited for this functionality 👍

Reply:
I think maybe we could make Go Template work. If we take the user input as a conditional, and insert it into a hard coded if statement?  e.g.

We can have a boilerplate template: 
```
{{ if <user expression> }}
ok
{{end}}
```

And when we evaluate a view expression, we replace `<user expression>` with the Go template expression defined by the user in the Corefile, and evaluate the template.  If it evaluates to the string "ok" then it's a pass.  This kind of feels like forcing the wrong puzzle piece.  Maybe there is a better way to use Go templates to do this - I don't have much experience with it.

The Go template expression syntax looks a bit unfriendly though:
* it appears to use prefix notation, which is not easy to read for most people.  
* function call syntax also doesn't appear to use parentheses to group all their arguments, so to read an expression that has a function call in it, you have to know how many arguments it takes.

I think those issues would be tolerable for very simple expressions that comprise a single function call, or a single comparison.

e.g. single operation no too bad
```
view incidr client_ip "1.2.3.0/24"
```

multiple operations ...  

```
view and eq class "IN" and eq type "A" incidr client_ip "1.2.3.0/24"
```

many operations ...
```
view and eq class "IN" or and eq type "A" incidr client_ip "1.2.3.0/24" and eq type "AAAA" incidr client_ip "1:2::3:4/64" 
```

What do you think?

Reply:
Hmm - looks like I could cut a bit less against the grain with Go templates for custom function calls, and attach methods to certain state fields. e.g. enabling cleaner syntax like:

```
view ClientIP.InCIDR "1.2.3.0/24"
```

But that doesn't really resolve the prefix notation thing.

The question I guess is, does the benefit of going with a Go-supported tool outweigh its apparently loose technical fit, and cryptic syntax.

Reply:
true, the GO template might not be the best fit for your use case, it would probably somehow work, but it seems it would be too hacky.

from what I have tried the `github.com/antonmedv/expr` seems like a nice fit. API seems quite understandable, also it seems to be well documented and actively developed